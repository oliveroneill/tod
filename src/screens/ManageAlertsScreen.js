import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  Dimensions,
  TouchableHighlight,
  ListView,
  ListViewDataSource,
  Alert,
  Button,
} from 'react-native';
import { List, ListItem } from 'react-native-elements';
import moment from 'moment';

import TouchableText from '../components/TouchableText.js';
import styles from '../style/Styles.js'
import Utils from '../utils/Utils.js'
import LoadingScreen from '../components/LoadingScreen.js';

let ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});

class ManageAlertsScreen extends Component {
  static navigationOptions = ({ navigation }) => {
    const {state, setParams, navigate} = navigation;
    const params = state.params || {};
    const editing = params.editing || false;
    const onEdit = params.onEdit;
    return {
      title: 'Manage Trips',
      headerRight: (
        <TouchableText
          style={{paddingRight: 10}}
          onPress={() => {
            setParams({editing: !editing})
            onEdit();
          }}
        >{editing ? "Done" : "Edit"}</TouchableText>
      ),
    };
  };
  state = {
    ds: ds,
    trips: [],
    editing: false,
    loading: false,
    errored: false
  };
  getTrips() {
    this.api.getScheduledTrips()
    .then((trips) => {
      this.setState({loading: false});
      this.setTrips(trips);
    })
    .catch((error) => {
      this.setState({errored: true});
    });
  }
  setTrips(trips) {
    this.setState({
      trips:trips,
      ds: this.state.ds.cloneWithRows(trips)
    })
  }
  enableDisableTrip(id) {
    this.setState({loading: true});
    this.api.enableDisableTrip(id)
    .then(() => {
      this.getTrips();
    });
  }
  deleteTrip(id) {
    this.setState({loading: true});
    this.api.deleteTrip(id)
    .then(() => {
      this.getTrips()
    });
  }
  onEdit() {
    this.setState({editing:!this.state.editing})
  }
  componentWillMount() {
    this.api = this.props.navigation.state.params.api;
    const { setParams } = this.props.navigation;
    setParams({onEdit: this.onEdit.bind(this)})
  }
  renderCell(rowData, sectionID) {
    let arrivalDate = moment.unix(rowData.route.arrival_time / 1000)
    let arrival = arrivalDate.format('h:mm a');
    let subtitle = "Arrive at "+arrival;
    let days = ["Mon", "Tue", "Wed", "Thurs", "Fri", "Sat", "Sun"];
    var repeats = "";
    for (var i = 0; i < rowData.repeat_days.length; i++) {
      if (rowData.repeat_days[i]) {
        if (repeats.length > 0) {
          repeats += ", ";
        }
        repeats += days[i]
      }
    }
    if (repeats.length > 0)
      subtitle += " Repeats: "+repeats;
    return (
      <ListItem
        key={sectionID}
        leftIcon={{name:Utils.getTransportIcon(rowData.transport_type)}}
        title={rowData.route.description}
        subtitle={subtitle}
        underlayColor='#e1e1e1'
        onPress={() => {Utils.openInMaps(arrivalDate, rowData.transport_type, rowData.origin, rowData.destination)}}
        rightIcon={{'name':'remove-circle', 'color': 'red'}}
        onPressRightIcon={() => this.deleteTrip(rowData.trip_id)}
        hideChevron={!this.state.editing}
        switchButton={!this.state.editing}
        onSwitch={() => this.enableDisableTrip(rowData.trip_id)}
        switched={rowData.enabled}
        subtitleNumberOfLines={3}
      />
    );
  }
  componentDidMount() {
    this.getTrips();
  }
  render() {
    return (
      <View style={styles.screen}>
        { this.state.loading ?
          <LoadingScreen
            loadingMessage="Loading Trips..."
            errored={this.state.errored}
            errorMessage="Something went wrong. Please ensure you have an internet connection."
            retry={this.getTrips.bind(this)}
          />
        :
          <View style={styles.container}>
            <Text style={{paddingLeft: 10, paddingTop:10}}>Trips: </Text>
            <List>
              <ListView
                enableEmptySections
                renderRow={this.renderCell.bind(this)}
                dataSource={this.state.ds}
              />
            </List>
          </View>
        }
      </View>
    )
  }
}

export default ManageAlertsScreen;
