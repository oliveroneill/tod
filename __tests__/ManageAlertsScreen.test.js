import React from 'react';
import renderer from 'react-test-renderer';
import { Switch, TouchableOpacity } from 'react-native';

import { shallow } from 'enzyme';

import ManageAlertsScreen from '../src/screens/ManageAlertsScreen.js';

describe('ManageAlertsScreen', () => {
  let mockRowData = {
    route: {
      arrival_time: 1482363467071,
      departure_time: 1482363367071,
      description: "Take x avenue",
      name: "Bus Route 1"
    },
    repeat_days: [false, true, true, false, false, false, true],
    transport_type:"walking",
    enabled: true,
    trip_id:"4543",
  };

  it('displays routes correctly', () => {
    let screen = new ManageAlertsScreen({navigation:jest.fn()});
    screen.state.editing = false;
    let component = renderer.create(screen.renderCell(mockRowData, 1));
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('displays routes correctly while editing', () => {
    let screen = new ManageAlertsScreen({navigation:jest.fn()});
    screen.state.editing = true;
    let component = renderer.create(screen.renderCell(mockRowData, 1));
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('handles enable', () => {
    let screen = new ManageAlertsScreen({navigation:jest.fn()});
    screen.api = {
      enableDisableTrip: jest.fn().mockImplementation(() => new Promise(function(resolve, reject) {
          resolve();
      }))
    };
    screen.setState = jest.fn();
    const wrapper = shallow(screen.renderCell(mockRowData, 1));
    wrapper.find(Switch).first().props().onValueChange();
    expect(screen.api.enableDisableTrip).toHaveBeenCalledWith(mockRowData.trip_id);
  });

  it('handles delete', () => {
    let screen = new ManageAlertsScreen({navigation:jest.fn()});
    screen.state.editing = true;
    screen.api = {
      deleteTrip: jest.fn().mockImplementation(() => new Promise(function(resolve, reject) {
          resolve();
      }))
    };
    screen.setState = jest.fn();
    const wrapper = shallow(screen.renderCell(mockRowData, 1));
    wrapper.find(TouchableOpacity).first().props().onPress();
    expect(screen.api.deleteTrip).toHaveBeenCalledWith(mockRowData.trip_id);
  });
});
