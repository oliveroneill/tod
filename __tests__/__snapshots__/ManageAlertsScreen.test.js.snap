// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageAlertsScreen displays routes correctly 1`] = `
<View
  onLongPress={undefined}
  onPress={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "#bbb",
        "borderBottomWidth": 0.5,
        "paddingBottom": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
      undefined,
    ]
  }
  underlayColor="#e1e1e1"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginLeft": 10,
        },
        undefined,
      ]
    }
  >
    <View
      onLongPress={undefined}
      onPress={undefined}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Object {
            "flex": 0.15,
          },
          undefined,
        ]
      }
      underlayColor="white"
    >
      <View>
        <View
          onPress={undefined}
          style={
            Array [
              false,
              false,
              false,
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
          underlayColor="white"
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#bdc6cf",
                  "fontSize": 24,
                },
                Array [
                  Object {
                    "backgroundColor": "transparent",
                  },
                  Array [
                    Object {
                      "marginRight": 8,
                    },
                    undefined,
                  ],
                ],
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "#43484d",
                  "fontSize": 17.5,
                },
                false,
                undefined,
                undefined,
              ],
            ]
          }
        >
          Take x avenue
        </Text>
      </View>
      <View
        style={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          numberOfLines={3}
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "#86939e",
                  "fontSize": 15,
                  "fontWeight": "600",
                  "marginTop": 1,
                },
                false,
                undefined,
                undefined,
              ],
            ]
          }
        >
          Arrive at 5:37 pm Repeats: Tue, Wed, Sun
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 0.15,
          "justifyContent": "center",
          "marginRight": 5,
        }
      }
    >
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor={undefined}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            undefined,
          ]
        }
        thumbTintColor={undefined}
        tintColor={undefined}
        value={true}
      />
    </View>
  </View>
</View>
`;

exports[`ManageAlertsScreen displays routes correctly while editing 1`] = `
<View
  onLongPress={undefined}
  onPress={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "#bbb",
        "borderBottomWidth": 0.5,
        "paddingBottom": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
      undefined,
    ]
  }
  underlayColor="#e1e1e1"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginLeft": 10,
        },
        undefined,
      ]
    }
  >
    <View
      onLongPress={undefined}
      onPress={undefined}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Object {
            "flex": 0.15,
          },
          undefined,
        ]
      }
      underlayColor="white"
    >
      <View>
        <View
          onPress={undefined}
          style={
            Array [
              false,
              false,
              false,
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
          underlayColor="white"
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#bdc6cf",
                  "fontSize": 24,
                },
                Array [
                  Object {
                    "backgroundColor": "transparent",
                  },
                  Array [
                    Object {
                      "marginRight": 8,
                    },
                    undefined,
                  ],
                ],
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "#43484d",
                  "fontSize": 17.5,
                },
                false,
                undefined,
                undefined,
              ],
            ]
          }
        >
          Take x avenue
        </Text>
      </View>
      <View
        style={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          numberOfLines={3}
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "#86939e",
                  "fontSize": 15,
                  "fontWeight": "600",
                  "marginTop": 1,
                },
                false,
                undefined,
                undefined,
              ],
            ]
          }
        >
          Arrive at 5:37 pm Repeats: Tue, Wed, Sun
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 0.15,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        onPress={undefined}
        style={
          Array [
            false,
            false,
            false,
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
        underlayColor="white"
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "red",
                "fontSize": 28,
              },
              Array [
                Object {
                  "backgroundColor": "transparent",
                },
                undefined,
              ],
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
